clinica/iotools/converters/adni_to_bids/adni_modalities/adni_fmap.py

```diff
@@ -245,165 +253,225 @@ def fmap_image(
+    filenames = [f for f in os.listdir(fmap_path) if not f.startswith(".")]
+    pattern = r"(.*_)fmap_(.*?)(\..*)$"
+    for previous_filename in filenames:
+        rgx = re.search(pattern, previous_filename)
+        cut, extension, type = rgx.group(1), rgx.group(2), rgx.group(3)
+        new_name = f"{cut}{phase_magnitude_renamer(extension, case)}{type}"
+        os.rename(fmap_path / previous_filename, fmap_path / new_name)
+
+
+def fmap_case_handler_factory(case: BIDSFMAPCase) -> Callable:
+    if case == BIDSFMAPCase.EMPTY_FOLDER:
+        return empty_folder_handler
+    if case == BIDSFMAPCase.ONE_PHASE_TWO_MAGNITUDES:
+        return one_phase_two_magnitudes_handler
+    if case == BIDSFMAPCase.TWO_PHASES_TWO_MAGNITUDES:
+        return two_phases_two_magnitudes_handler
+    if case == BIDSFMAPCase.DIRECT_FIELDMAPS:
+        return direct_fieldmaps_handler
+    if case == BIDSFMAPCase.NOT_SUPPORTED:
+        return not_supported_handler
+
+
+def get_json_file_matching_pattern(fmap_path: Path, pattern: str) -> Path:
+    json_file = [
+        f
+        for f in fmap_path.iterdir()
+        if not f.name.startswith(".") and f.name.endswith(pattern)
+    ]
+    if len(json_file) != 1:
+        msg = f"Expected only a single JSON file ending in '{pattern}' in {fmap_path}, but got:"
+        msg += "\n".join([f.name for f in json_file])
+        raise ValueError(msg)
+    return json_file[0]
+
+
+def empty_folder_handler(fmap_path: Path):
+    sub = fmap_path.parent.parent.name
+    ses = fmap_path.parent.name
+    cprint(f"Folder for {sub}, {ses} is empty", lvl="warning")
+
+
+def one_phase_two_magnitudes_handler(fmap_path: Path):
+    "Performs the checks and renaming associated to BIDS specifications, case 1 for fieldmaps"
+    sub = fmap_path.parent.parent.name
+    ses = fmap_path.parent.name
+
+    cprint(
+        f"BIDS Case 1 : expecting 1 phasediff and at least 1 magnitude files for {sub}, {ses}.",
+        lvl="info",
+    )
+
+    files = [f for f in os.listdir(fmap_path) if not f.startswith(".")]
+    json_file = get_json_file_matching_pattern(fmap_path, pattern="ph.json")
+
+    # Checking for Echotime keys in phase json
+    with open(fmap_path / json_file, "r") as file:
+        json_data = json.load(file)
+    if "EchoTime1" not in json_data or "EchoTime2" not in json_data:
+        cprint(f'Invalid file {json_file}, missing "EchoTime" key.', lvl="warning")
+        unrecognized_fmap_case(fmap_path)
+        return
+    rename_files(fmap_path, BIDSFMAPCase.ONE_PHASE_TWO_MAGNITUDES)
+
+
+def two_phases_two_magnitudes_handler(fmap_path: Path):
+    """Performs the checks and renaming for BIDS spec case 2 for fieldmaps"""
+
+    sub = fmap_path.parent.parent.name
+    ses = fmap_path.parent.name
+    cprint(
+        f"BIDS Case 2 : expecting 2 phase and 2 magnitude files for {sub}, {ses}.",
+        lvl="info",
+    )
+
+    files = [f for f in os.listdir(fmap_path) if not f.startswith(".")]
```

Comment:
Let's remove this last bit of redundancy:

```python
def check_json_contains_keys(json_file: Path, keys: Iterable[str]) -> bool:
    with open(json_file, "r") as file:
        json_data = json.load(file)
    missing_keys = [key for key in keys if key not in json_data]
    if missing_keys:
        cprint(f'Invalid file {json_file}, missing the following keys: {missing_keys}.', lvl="warning")
        return False
    return True
```

Here, would be used like:

```python
if not all([check_json_contains_keys(f, "EchoTime") for f in check_json]):
    unrecognized_fmap_case(fmap_path)
else:
    rename_files(fmap_path, BIDSFMAPCase.TWO_PHASES_TWO_MAGNITUDES)
```

In `one_phase_two_magnitudes_handler`:

```python
if not check_json_contains_keys(json_file, ("EchoTime1", "EchoTime2")):
    unrecognized...
else:
    ...
```